File: apps/api/src/kick_logs/infrastructure/kick/event_parser.py
import json
from dataclasses import dataclass
from typing import Any


@dataclass(frozen=True, slots=True)
class KickChatMessageEvent:
    event: str
    channel: str | None
    payload: dict[str, Any]


class KickEventParser:
    chat_message_event = r"App\Events\ChatMessageEvent"

    def parse(self, raw_event: str | dict[str, Any]) -> KickChatMessageEvent | None:
        envelope = self._read_json_object(raw_event)
        if envelope is None:
            return None

        if envelope.get("event") != self.chat_message_event:
            return None

        payload = self._read_json_object(envelope.get("data"))
        if payload is None or not self._has_required_message_fields(payload):
            return None

        return KickChatMessageEvent(
            event=str(envelope["event"]),
            channel=self._clean_text(envelope.get("channel")),
            payload=payload,
        )

    def _read_json_object(self, value: str | dict[str, Any] | Any) -> dict[str, Any] | None:
        if isinstance(value, dict):
            return value
        if not isinstance(value, str) or not value.strip():
            return None
        try:
            parsed = json.loads(value)
        except json.JSONDecodeError:
            return None
        return parsed if isinstance(parsed, dict) else None

    def _has_required_message_fields(self, payload: dict[str, Any]) -> bool:
        sender = payload.get("sender")
        return (
            self._clean_text(payload.get("id")) is not None
            and payload.get("chatroom_id") is not None
            and "content" in payload
            and isinstance(sender, dict)
            and self._clean_text(sender.get("id")) is not None
            and self._clean_text(sender.get("username")) is not None
        )

    def _clean_text(self, value: Any) -> str | None:
        if value is None:
            return None
        cleaned = str(value).strip()
        return cleaned or None

File: apps/api-go/internal/usecase/listener/event_parser.go
package listener

import (
	"encoding/json"
	"strconv"
	"strings"
)

const chatMessageEventName = `App\Events\ChatMessageEvent`

type ChatMessageEvent struct {
	EventName     string
	PusherChannel string
	Payload       map[string]any
	RawJSON       string
}

type EventParser struct{}

func NewEventParser() EventParser {
	return EventParser{}
}

func (EventParser) Parse(raw string) (ChatMessageEvent, bool) {
	envelope, ok := readJSONObject(raw)
	if !ok {
		return ChatMessageEvent{}, false
	}
	eventName, ok := envelope["event"].(string)
	if !ok || eventName != chatMessageEventName {
		return ChatMessageEvent{}, false
	}
	payload, ok := readJSONValueObject(envelope["data"])
	if !ok {
		return ChatMessageEvent{}, false
	}

	channel, _ := envelope["channel"].(string)
	return ChatMessageEvent{
		EventName:     eventName,
		PusherChannel: strings.TrimSpace(channel),
		Payload:       payload,
		RawJSON:       raw,
	}, true
}

func readJSONObject(raw string) (map[string]any, bool) {
	if strings.TrimSpace(raw) == "" {
		return nil, false
	}
	var value map[string]any
	if err := json.Unmarshal([]byte(raw), &value); err != nil {
		return nil, false
	}
	return value, true
}

func readJSONValueObject(value any) (map[string]any, bool) {
	if object, ok := value.(map[string]any); ok {
		return object, true
	}
	text, ok := value.(string)
	if !ok {
		return nil, false
	}
	return readJSONObject(text)
}

func hasRequiredMessageFields(payload map[string]any) bool {
	sender, ok := payload["sender"].(map[string]any)
	if !ok {
		return false
	}
	return cleanText(payload["id"]) != "" &&
		asInt64(payload["chatroom_id"]) > 0 &&
		payloadHasContent(payload) &&
		asInt64(sender["id"]) > 0 &&
		cleanText(sender["username"]) != ""
}

func payloadHasContent(payload map[string]any) bool {
	_, ok := payload["content"]
	return ok
}

func chatroomIDFromPusherChannel(channel string) int64 {
	parts := strings.Split(strings.TrimSpace(channel), ".")
	if len(parts) < 2 || parts[0] != "chatrooms" {
		return 0
	}
	value, _ := strconv.ParseInt(parts[1], 10, 64)
	return value
}
